src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2998,5 +2998,92 @@ public static bool IsFunctionPointerCallingConventionContext(this SyntaxTree syn
                    targetToken.Parent is FunctionPointerTypeSyntax functionPointerType &&
                    targetToken == functionPointerType.AsteriskToken;
         }
+
+        /// <summary>
+        /// Tells if in a position of a base list, where class name is a valid completion.
+        /// For instatnce, this flag will be <see langword="true" /> for
+        /// <code>class C : $$</code>
+        /// but not for
+        /// <code>class C : A, $$</code>
+        /// </summary>
+        public static bool IsBaseClassContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            var possibleBaseListSyntax = targetToken.GetAncestor<BaseListSyntax>();
+            var possibleBaseTypeSyntax = targetToken.GetAncestor<BaseTypeSyntax>();
+
+            if (possibleBaseListSyntax?.Parent is not ClassDeclarationSyntax)
+            {
+                return false;
+            }
+
+            // We are in a non-written base type, e.g. "class C : $$"
+            if (possibleBaseTypeSyntax is null)
+            {
+                return targetToken.IsKind(SyntaxKind.ColonToken);
+            }
+
+            // 1. Current base type must be the first in the base list, because class inheritance is valid only at the first position in the base list
+            // 2. The only available target token type is a qualified name or an alias qualified name. If there is generic names or whatever else,
+            // then it is not directly inheritance context and our restrictions are not valid there
+            return possibleBaseListSyntax.Types[0].Equals(possibleBaseTypeSyntax) &&
+                   targetToken.Parent is QualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } }
+                                      or AliasQualifiedNameSyntax { Parent: BaseTypeSyntax or QualifiedNameSyntax { Parent: BaseTypeSyntax } };
```

Comment:
what if you have:  `List<x.$$Y.Z.W.M>`  the targettoken parent will be a QNS.  But hte parent, or parent/parent won't be the basetypesyntax.

Reply:
same with `X.$$Y.Z.W.M`.